### PR TITLE
Synchronize runtime id across versions of the tracer

### PR DESCRIPTION
**Pull request by kevingosse**
- Add a distributed tracer API to synchronize the runtime id across versions of the tracer
- When multiple versions of the tracer are loaded, only enable runtime metrics in the automatic one
- Remove the possibility to disable runtime metrics from code, as synchronizing that state across versions would greatly increase the complexity of the change

Changes in public API:

```diff
    public class TracerSettings
    {
        public const string DefaultAgentHost = "localhost";
        public const int DefaultAgentPort = 8126;
        public TracerSettings() { }
        public TracerSettings(Datadog.Trace.Configuration.IConfigurationSource source) { }
        public System.Uri AgentUri { get; set; }
        [System.Obsolete("App Analytics has been replaced by Tracing without Limits. For more information s" +
            "ee https://docs.datadoghq.com/tracing/legacy_app_analytics/")]
        public bool AnalyticsEnabled { get; set; }
        public string CustomSamplingRules { get; set; }
        public bool DiagnosticSourceEnabled { get; set; }
        public System.Collections.Generic.HashSet<string> DisabledIntegrationNames { get; set; }
        public int DogStatsdPort { get; set; }
        public string Environment { get; set; }
        public double? GlobalSamplingRate { get; set; }
        public System.Collections.Generic.IDictionary<string, string> GlobalTags { get; set; }
        public System.Collections.Generic.IDictionary<string, string> HeaderTags { get; set; }
        public Datadog.Trace.Configuration.IntegrationSettingsCollection Integrations { get; }
        public bool KafkaCreateConsumerScopeEnabled { get; set; }
        public bool LogsInjectionEnabled { get; set; }
        public int MaxTracesSubmittedPerSecond { get; set; }
        public string MetricsPipeName { get; set; }
        public bool PartialFlushEnabled { get; set; }
        public int PartialFlushMinSpans { get; set; }
-       public bool RuntimeMetricsEnabled { get; set; }
        public string ServiceName { get; set; }
        public string ServiceVersion { get; set; }
        public bool StartupDiagnosticLogEnabled { get; set; }
        public bool TraceEnabled { get; set; }
        public bool TracerMetricsEnabled { get; set; }
        public string TracesPipeName { get; set; }
        public int TracesPipeTimeoutMs { get; set; }
        public string TracesTransport { get; set; }
        public Datadog.Trace.Configuration.ImmutableTracerSettings Build() { }
        public void SetHttpClientErrorStatusCodes(System.Collections.Generic.IEnumerable<int> statusCodes) { }
        public void SetHttpServerErrorStatusCodes(System.Collections.Generic.IEnumerable<int> statusCodes) { }
        public void SetServiceNameMappings(System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, string>> mappings) { }
        public static Datadog.Trace.Configuration.CompositeConfigurationSource CreateDefaultConfigurationSource() { }
        public static Datadog.Trace.Configuration.TracerSettings FromDefaultSources() { }
    }

    public class ImmutableTracerSettings
    {
        public ImmutableTracerSettings(Datadog.Trace.Configuration.IConfigurationSource source) { }
        public ImmutableTracerSettings(Datadog.Trace.Configuration.TracerSettings settings) { }
        public System.Uri AgentUri { get; }
        [System.Obsolete("App Analytics has been replaced by Tracing without Limits. For more information s" +
            "ee https://docs.datadoghq.com/tracing/legacy_app_analytics/")]
        public bool AnalyticsEnabled { get; }
        public string CustomSamplingRules { get; }
        public int DogStatsdPort { get; }
        public string Environment { get; }
        public double? GlobalSamplingRate { get; }
        public System.Collections.Generic.IReadOnlyDictionary<string, string> GlobalTags { get; }
        public System.Collections.Generic.IReadOnlyDictionary<string, string> HeaderTags { get; }
        public Datadog.Trace.Configuration.ImmutableIntegrationSettingsCollection Integrations { get; }
        public bool KafkaCreateConsumerScopeEnabled { get; }
        public bool LogsInjectionEnabled { get; }
        public int MaxTracesSubmittedPerSecond { get; }
        public string MetricsPipeName { get; }
        public bool PartialFlushEnabled { get; }
        public int PartialFlushMinSpans { get; }
-       public bool RuntimeMetricsEnabled { get; }
        public string ServiceName { get; }
        public string ServiceVersion { get; }
        public bool StartupDiagnosticLogEnabled { get; }
        public bool TraceEnabled { get; }
        public bool TracerMetricsEnabled { get; }
        public string TracesPipeName { get; }
        public int TracesPipeTimeoutMs { get; }
        public string TracesTransport { get; }
        public static Datadog.Trace.Configuration.ImmutableTracerSettings FromDefaultSources() { }
    }
```

### CODE REVIEW

**Review by pierotibou — APPROVED** [review]

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2172 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) will **not change** line coverage
:heavy_check_mark: Merging #2172 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) will **not change** branch coverage
:heavy_check_mark: Merging #2172 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) will will **decrease** complexity by `39`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) | #2172       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9142` / `12698` | `9067` / `12613` |          |
| Lines %   | `72%`      | `72%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4135` / `6193` | `4109` / `6165` |          |
| Branches %| `67%`      | `67%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7365`      | `7326`        |  `-39`  :heavy_check_mark:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=83814&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/83814/artifacts?artifactName=Code%20Coverage%20Report_83814&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2172](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=83895&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/83895/artifacts?artifactName=Code%20Coverage%20Report_83895&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) | #2172       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `72%`      | `72%`       |  `0%` :heavy_check_mark:  |
| Branches %| `67%`      | `67%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7365`      | `7326`       |  `-39` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Couchbase.IIOServiceExecuteIntegrationBis](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Couchbase/IIOServiceExecuteIntegrationBis.cs) | `-100%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2.TestMethodRunnerExecuteIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodRunnerExecuteIntegration.cs) | `-33%` :no_entry: | `-19%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Ci.Agent.CIAgentWriter](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs) | `-10%` :no_entry: | `-21%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.NUnit.NUnitCompositeWorkItemSkipChildrenIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/NUnit/NUnitCompositeWorkItemSkipChildrenIntegration.cs) | `-8%` :no_entry: | `-8%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.ManualTracer](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/ClrProfiler/ManualTracer.cs) | `0%` :heavy_check_mark: | `-8%` :no_entry:   | `1` :no_entry: |
| [Datadog.Trace.ClrProfiler.AutomaticTracer](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/ClrProfiler/AutomaticTracer.cs) | `0%` :heavy_check_mark: | `8%` :heavy_check_mark:   | `3` :no_entry: |
| [Datadog.Trace.ClrProfiler.CommonTracer](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/ClrProfiler/CommonTracer.cs) | `7%` :heavy_check_mark: | `-3%` :warning:   | `-3` :heavy_check_mark: |
| [Datadog.Trace.Logging.LibLogScopeEventSubscriber](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/Logging/LogsInjection/LibLogScopeEventSubscriber.cs) | `8%` :heavy_check_mark: | `1%` :heavy_check_mark:   | `-6` :heavy_check_mark: |
| [Datadog.Trace.Logging.NoOpSerilogLogProvider](https://github.com/DataDog/dd-trace-dotnet/tree/ad50683458a039a502d2b37c8969503fa46d14c0/tracer/src/Datadog.Trace/Logging/LogsInjection/NoOpSerilogLogProvider.cs) | `86%` :heavy_check_mark: | `75%` :heavy_check_mark:   | `-3` :heavy_check_mark: |


The following classes were added in #2172:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.Serilog.LogsInjection.LoggerDispatchInstrumentation](https://github.com/DataDog/dd-trace-dotnet/pull/2172/files) | `0%` | `0%` | `6` |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.Serilog.LogsInjection.SerilogLogPropertyHelper\`1](https://github.com/DataDog/dd-trace-dotnet/pull/2172/files) | `100%` | `100%` | `2` |


2 classes were removed from Datadog.Trace in #2172


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=83814&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/83814/artifacts?artifactName=Code%20Coverage%20Report_83814&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2172](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=83895&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/83895/artifacts?artifactName=Code%20Coverage%20Report_83895&api-version=6.0&%24format=zip)

File: tracer/test/Datadog.Trace.Tests/DistributedTracer/AutomaticTracerTests.cs [hunk 1]
```diff
@@ -96,5 +97,19 @@ public void GetDistributedTrace()
 
             automaticTracer.GetDistributedTrace().Should().BeNull();
         }
+
+        [Fact]
+        public void RuntimeId()
```
Comment (zacharycmontoya) [inline tracer/test/Datadog.Trace.Tests/DistributedTracer/AutomaticTracerTests.cs:102]: Similar to `ManualTracerTests.IsChildTracer`, can you add a `AutomaticTracerTests.IsNotChildTracer` test?

**Review by zacharycmontoya — APPROVED** [review]
Nice work 👍🏼

**Review by lucaspimentel — APPROVED** [review]

File: tracer/test/Datadog.Trace.Tests/DistributedTracer/AutomaticTracerTests.cs (hunk 1 above)
Comment (kevingosse) [reply tracer/test/Datadog.Trace.Tests/DistributedTracer/AutomaticTracerTests.cs:102]: Good catch, added in https://github.com/DataDog/dd-trace-dotnet/pull/2172/commits/2190b6fe2e868c3f09fd9f69ea88429ec0d5d200

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2172 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) will **not change** line coverage
:heavy_check_mark: Merging #2172 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) will **not change** branch coverage
:heavy_check_mark: Merging #2172 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) will will **decrease** complexity by `21`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) | #2172       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9142` / `12698` | `9085` / `12636` |          |
| Lines %   | `72%`      | `72%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4135` / `6193` | `4120` / `6179` |          |
| Branches %| `67%`      | `67%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7365`      | `7344`        |  `-21`  :heavy_check_mark:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=83814&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/83814/artifacts?artifactName=Code%20Coverage%20Report_83814&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2172](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84007&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84007/artifacts?artifactName=Code%20Coverage%20Report_84007&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c378bdba5002f32f7ff4fe1beb5a23beb8da59c9) | #2172       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `72%`      | `72%`       |  `0%` :heavy_check_mark:  |
| Branches %| `67%`      | `67%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7365`      | `7344`       |  `-21` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Couchbase.IIOServiceExecuteIntegrationBis](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Couchbase/IIOServiceExecuteIntegrationBis.cs) | `-100%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.MsTestV2.TestMethodRunnerExecuteIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodRunnerExecuteIntegration.cs) | `-33%` :no_entry: | `-19%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Ci.Agent.CIAgentWriter](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs) | `-10%` :no_entry: | `-21%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Testing.NUnit.NUnitCompositeWorkItemSkipChildrenIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/NUnit/NUnitCompositeWorkItemSkipChildrenIntegration.cs) | `-8%` :no_entry: | `-8%` :no_entry:   | `0` :heavy_check_mark: |
| [Datadog.Trace.Agent.TransportStrategy](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/Agent/TransportStrategy.cs) | `0%` :heavy_check_mark: | `-8%` :no_entry:   | `-2` :heavy_check_mark: |
| [Datadog.Trace.ClrProfiler.ManualTracer](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/ClrProfiler/ManualTracer.cs) | `0%` :heavy_check_mark: | `-8%` :no_entry:   | `1` :no_entry: |
| [Datadog.Trace.ClrProfiler.AutomaticTracer](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/ClrProfiler/AutomaticTracer.cs) | `0%` :heavy_check_mark: | `8%` :heavy_check_mark:   | `3` :no_entry: |
| [Datadog.Trace.ClrProfiler.CommonTracer](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/ClrProfiler/CommonTracer.cs) | `7%` :heavy_check_mark: | `-3%` :warning:   | `-3` :heavy_check_mark: |
| [Datadog.Trace.Logging.LibLogScopeEventSubscriber](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/Logging/LogsInjection/LibLogScopeEventSubscriber.cs) | `8%` :heavy_check_mark: | `1%` :heavy_check_mark:   | `-6` :heavy_check_mark: |
| [Datadog.Trace.Logging.NoOpSerilogLogProvider](https://github.com/DataDog/dd-trace-dotnet/tree/60da684cd2e738c222909f7c0b24986a2151c659/tracer/src/Datadog.Trace/Logging/LogsInjection/NoOpSerilogLogProvider.cs) | `86%` :heavy_check_mark: | `75%` :heavy_check_mark:   | `-3` :heavy_check_mark: |


The following classes were added in #2172:

| File    | Line coverage | Branch coverage | Complexity |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.Serilog.LogsInjection.LoggerDispatchInstrumentation](https://github.com/DataDog/dd-trace-dotnet/pull/2172/files) | `0%` | `0%` | `6` |
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.Serilog.LogsInjection.SerilogLogPropertyHelper\`1](https://github.com/DataDog/dd-trace-dotnet/pull/2172/files) | `100%` | `100%` | `2` |
| [Datadog.Trace.Configuration.ExporterSettings](https://github.com/DataDog/dd-trace-dotnet/pull/2172/files) | `91%` | `87%` | `56` |
| [Datadog.Trace.Configuration.ImmutableExporterSettings](https://github.com/DataDog/dd-trace-dotnet/pull/2172/files) | `75%` | `100%` | `2` |


2 classes were removed from Datadog.Trace in #2172


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=83814&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/83814/artifacts?artifactName=Code%20Coverage%20Report_83814&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2172](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=84007&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/84007/artifacts?artifactName=Code%20Coverage%20Report_84007&api-version=6.0&%24format=zip)

**Pull request merged**